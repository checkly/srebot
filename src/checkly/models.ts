export type ErrorMessage = {
  message: string;
  stack: string;
};
export type LogEntry = {
  time: number;
  msg: string;
  level: string;
};

export class CheckGroup {
  id: number;
  name: string;
  concurrency: number;
  apiCheckDefaults: {
    url: string;
    headers: {
      key: string;
      value: string;
      locked: boolean;
    }[];
    basicAuth: {
      password: string;
      username: string;
    };
    assertions: any[];
    queryParameters: any[];
  };
  alertSettings: {
    reminders: {
      amount: number;
      interval: number;
    };
    escalationType: string;
    runBasedEscalation: {
      failedRunThreshold: number;
    };
    timeBasedEscalation: {
      minutesFailingThreshold: number;
    };
  };
  environmentVariables: any[];
  setupSnippetId: string | null;
  tearDownSnippetId: string | null;
  localSetupScript: string | null;
  localTearDownScript: string | null;
  activated: boolean;
  muted: boolean;
  useGlobalAlertSettings: boolean;
  doubleCheck: boolean;
  locations: string[];
  tags: any[];
  created_at: string;
  updated_at: string;
  runtimeId: string | null;
  retryStrategy: string | null;
  runParallel: boolean;
  alertChannelSubscriptions: {
    alertChannelId: number;
    activated: boolean;
  }[];
  privateLocations: any[];
}

export class Check {
  id: string;
  checkType: string;
  name: string;
  frequency: number;
  frequencyOffset: number;
  activated: boolean;
  muted: boolean;
  shouldFail: boolean;
  locations: string[];
  script: string | null;
  created_at: string;
  updated_at: string;
  environmentVariables: any[];
  doubleCheck: boolean;
  tags: any[];
  sslCheckDomain: string;
  setupSnippetId: string | null;
  tearDownSnippetId: string | null;
  localSetupScript: string | null;
  localTearDownScript: string | null;
  alertSettings: {
    reminders: {
      amount: number;
      interval: number;
    };
    escalationType: string;
    runBasedEscalation: {
      failedRunThreshold: number;
    };
    timeBasedEscalation: {
      minutesFailingThreshold: number;
    };
  };
  useGlobalAlertSettings: boolean;
  degradedResponseTime: number;
  maxResponseTime: number;
  groupId: number | null;
  groupOrder: number | null;
  runtimeId: string | null;
  scriptPath: string | null;
  retryStrategy: string | null;
  runParallel: boolean;
  request: {
    method: string;
    url: string;
    body: string;
    bodyType: string;
    headers: {
      key: string;
      value: string;
      locked: boolean;
    }[];
    queryParameters: any[];
    assertions: any[];
    basicAuth: {
      password: string;
      username: string;
    };
    followRedirects: boolean;
    skipSSL: boolean;
    ipFamily: string;
  };
  alertChannelSubscriptions: {
    alertChannelId: number;
    activated: boolean;
  }[];
  privateLocations: any[];
}

export class CheckResult {
  id: string;
  name: string;
  checkId: string;
  hasFailures: boolean;
  hasErrors: boolean;
  isDegraded: boolean;
  overMaxResponseTime: boolean;
  runLocation: string;
  startedAt: string; // ISO date string
  stoppedAt: string; // ISO date string
  created_at: string; // ISO date string
  responseTime: number;
  apiCheckResult: {
    assertions: Array<{
      source: string;
      target: number;
    }>;
    request: {
      method: string;
      url: string;
      data: string;
      headers: Record<string, string>;
      params: Record<string, string>;
    };
    response: {
      status: number;
      statusText: string;
      body: string;
      headers: Record<string, string>;
      timings: Record<string, unknown>;
      timingPhases: Record<string, unknown>;
    };
    requestError: string | null;
    jobLog: {
      setup: Array<LogEntry>;
      request: Array<LogEntry>;
      teardown: Array<LogEntry>;
    };
    jobAssets: unknown | null;
  } | null;
  browserCheckResult: {
    jobLog: Array<LogEntry>;
    playwrightTestTraces: Array<string>;
  } | null;
  multiStepCheckResult: {
    errors: Array<ErrorMessage>;
    endTime: number;
    startTime: number;
    runtimeVersion: string;
    jobLog: Array<LogEntry>;
    jobAssets: Array<unknown>;
    playwrightTestTraces: Array<string>;
    playwrightTestJsonReportFile: string;
  } | null;
  checkRunId: number;
  attempts: number;
  resultType: string;
  sequenceId: string;

  getLog(): string {
    const jobLog =
      this.apiCheckResult?.jobLog.request ||
      this.browserCheckResult?.jobLog ||
      this.multiStepCheckResult?.jobLog ||
      [];

    return jobLog
      .map(
        (logEntry) => `${logEntry.time} - ${logEntry.level}: ${logEntry.msg}`
      )
      .join("\n");
  }
}
