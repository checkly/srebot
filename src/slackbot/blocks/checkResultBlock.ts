import { Check, CheckResult } from "../../checkly/models";
import { SummarizeErrorsPromptType } from "../../prompts/checkly";

export interface CheckResultBlockProps {
  check: Check;
  checkAppUrl: string;
  checkResult: CheckResult;
  checkResultAppUrl: string;
  errorGroups: SummarizeErrorsPromptType;
  failingCheckResults: CheckResult[];
  intervalStart: Date;
}

export function createCheckResultBlock({
  check,
  checkAppUrl,
  checkResult,
  checkResultAppUrl,
  errorGroups,
  failingCheckResults,
  intervalStart,
}: CheckResultBlockProps) {
  const errorGroup = errorGroups.groups.find(
    (g) => g.checkResults.indexOf(checkResult.id) > -1,
  );

  return {
    text: `*Check Result Details*`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: "Check Result Details",
          emoji: true,
        },
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*Check Name:*\n<${checkAppUrl}|${check.name}>`,
          },
          {
            type: "mrkdwn",
            text: `*Timestamp:*\n${intervalStart.toISOString()}`,
          },
          {
            type: "mrkdwn",
            text: `*Location:*\n\`${checkResult.runLocation}\``,
          },
          {
            type: "mrkdwn",
            text: `*Check Result:*\n<${checkResultAppUrl}|Link>`,
          },
        ],
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Error Message:*\n\`${errorGroup?.errorMessage}\`\n\n*Similar Failures:*\nThis error occurred \`${errorGroup?.checkResults.length} times\` across locations: \`${check.locations.join("\`, \`")}\`. We saw \`${errorGroups.groups.length} Error Groups\` in total with \`${failingCheckResults.length} failures\` in the same timeframe.`,
        },
      },
    ],
  };
}
