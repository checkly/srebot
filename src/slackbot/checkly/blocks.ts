import {
  CHECKLY_ANALYZE_CHECK_ACTION,
  CHECKLY_LIST_CHECKS_ACTIONS,
  CHECKLY_RUN_CHECK_ACTION,
} from "./slack";

export const summaryBlock = (
  failed: number,
  degraded: number,
  passed: number,
) => {
  return {
    text: `Checkly Production Status: ${failed} Failed | ${degraded} Degraded | ${passed} Passed`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: "Checkly Production Status",
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `:x: *${failed}* Failed  |  :warning: *${degraded}* Degraded  |  :white_check_mark: *${passed}* Passed\n`,
        },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Failing Checks",
              emoji: true,
            },
            action_id: CHECKLY_LIST_CHECKS_ACTIONS.FAILING,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Degraded Checks",
              emoji: true,
            },
            action_id: CHECKLY_LIST_CHECKS_ACTIONS.DEGRADED,
          },
        ],
      },
    ],
  };
};

export const checkAttachment = ({
  checkId,
  check,
  group,
  type,
  tags,
  success24h,
  success7d,
  timestamp,
  location,
  executionTime,
  executionTimeAvg,
  executionTimeP95,
  executionTimeP99,
}: {
  checkId: string;
  check: string;
  group: string;
  type: string;
  tags: string[];
  success24h: number;
  success7d: number;
  timestamp: Date;
  location: string;
  executionTime: number;
  executionTimeAvg: number;
  executionTimeP95: number;
  executionTimeP99: number;
}) => {
  return {
    blocks: [
      {
        type: "divider",
      },
      {
        type: "header",
        text: {
          type: "plain_text",
          text: check,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Type:* ${type}\n*Group:* ${group}\n*Tags*: ${tags.join(", ")}`,
        },
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*When*\n<!date^${Math.floor(timestamp.getTime() / 1000)}^{ago}|${timestamp}>`,
          },
          {
            type: "mrkdwn",
            text: "*Region*\n" + location,
          },
        ],
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*Execution Time*\n${executionTime}s (${executionTimeAvg}s Avg | ${executionTimeP95}s P95 | ${executionTimeP99}s P99)`,
          },
          {
            type: "mrkdwn",
            text: `*Reliability*\n${success24h}% in 24h | ${success7d}% in 7d`,
          },
        ],
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "View Check",
              emoji: true,
            },
            url: `https://app.checklyhq.com/checks/${checkId}`,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Run Check",
              emoji: true,
            },
            value: checkId,
            action_id: CHECKLY_RUN_CHECK_ACTION,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Analyze",
              emoji: true,
            },
            value: checkId,
            action_id: CHECKLY_ANALYZE_CHECK_ACTION,
          },
        ],
      },
    ],
  };
};
