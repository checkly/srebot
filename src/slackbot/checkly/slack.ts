import { PrometheusCheckMetric, Reporting, Status } from "../../checkly/models";
import { checkly } from "../../checkly/client";
import { checkAttachment, summaryBlock } from "./blocks";
import { featureCoveragePrompt } from "../../prompts/checkly";
import { generateText } from "ai";
import { text } from "express";

export const CHECKLY_COMMAN_NAME = "/checkly";

export const checklyCommandHandler = async ({ ack, respond }) => {
  await ack();

  const metrics = await checkly.getPrometheusCheckStatus();

  const failing = metrics.failing.length;
  const degraded = metrics.degraded.length;
  const passing = metrics.passing.length;

  const { text, blocks } = summaryBlock(failing, degraded, passing);

  await respond({
    text,
    blocks,
  });
};

export const CHECKLY_LIST_CHECKS_ACTION =
  /checkly-list-checks-(failing|degraded|passing)/;

export enum CHECKLY_LIST_CHECKS_ACTIONS {
  FAILING = "checkly-list-checks-failing",
  DEGRADED = "checkly-list-checks-degraded",
  PASSING = "checkly-list-checks-passing",
}

export const checklyListChecksActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  const [rep24h, rep7d, statuses, metrics] = await Promise.all([
    checkly.getReporting({
      quickRange: "last24Hrs",
    }),
    checkly.getReporting({ quickRange: "last7Days" }),
    checkly.getStatuses(),
    checkly.getPrometheusCheckStatus(),
  ]);

  const rep24hMap = rep24h.reduce(
    reduceReporting,
    {} as Record<string, Reporting>,
  );
  const rep7dMap = rep7d.reduce(
    reduceReporting,
    {} as Record<string, Reporting>,
  );

  const statusesMap = statuses.reduce(
    reduceStatus,
    {} as Record<string, Status>,
  );

  const filter = action.action_id as CHECKLY_LIST_CHECKS_ACTIONS;
  let checksToShow: any[];
  let attachmentColor: string;

  switch (filter) {
    case CHECKLY_LIST_CHECKS_ACTIONS.FAILING:
      checksToShow = metrics.failing;
      attachmentColor = "#ff0000"; // Fixed syntax error - changed : to =
      break;
    case CHECKLY_LIST_CHECKS_ACTIONS.DEGRADED:
      checksToShow = metrics.degraded;
      attachmentColor = "#ffa500";
      break;
    case CHECKLY_LIST_CHECKS_ACTIONS.PASSING:
      checksToShow = metrics.passing;
      attachmentColor = "#36a64f";
      break;
    default:
      checksToShow = [];
      attachmentColor = "";
  }

  const attachments = checksToShow
    .map(PrometheusCheckMetric.fromJson)
    .map((m) => ({
      checkRunId: statusesMap[m.checkId].lastCheckRunId,
      checkId: m.checkId,
      check: m.name,
      group: m.group,
      type: m.checkType,
      tags: m.tags,
      success24h: rep24hMap[m.checkId].aggregate.successRatio,
      success7d: rep7dMap[m.checkId].aggregate.successRatio,
      timestamp: new Date(statusesMap[m.checkId].updated_at),
      location: statusesMap[m.checkId].lastRunLocation || "Unknown",
      executionTime: rep7dMap[m.checkId].aggregate.avg,
      executionTimeAvg: rep7dMap[m.checkId].aggregate.avg,
      executionTimeP95: rep7dMap[m.checkId].aggregate.p95,
      executionTimeP99: rep7dMap[m.checkId].aggregate.p99,
    }))
    .map((m) => ({
      color: attachmentColor,
      blocks: checkAttachment(m).blocks,
    }));

  await respond({
    text: "Checks:",
    attachments,
  });

  return;
};

const reduceReporting = (
  acc: Record<string, Reporting>,
  curr: Reporting,
): Record<string, Reporting> => {
  acc[curr.checkId] = curr;
  return acc;
};

const reduceStatus = (
  acc: Record<string, Status>,
  curr: Status,
): Record<string, Status> => {
  acc[curr.checkId] = curr;
  return acc;
};

export const CHECKLY_RUN_CHECK_ACTION = "checkly-run-check";

export const checklyRunCheckActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  await checkly.runCheck(JSON.parse((action as any).value).checkId);

  await respond("Check triggered!");
};

export const CHECKLY_ANALYZE_CHECK_ACTION = "checkly-analyze-check";

export const checklyAnalyzeCheckActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  const checkId = action.value;

  const check = await checkly.getCheck(checkId);
  const lastFailingCheckResult = await checkly.getCheckResults(
    checkId,
    true,
    1,
  );

  if (lastFailingCheckResult.length == 0) {
    await respond(
      `Unable to find failure for check ${checkId}.\n<https://app.checklyhq.com/checks/${checkId}|Open the check in your Browser instead.>`,
    );
    return;
  }

  if (!lastFailingCheckResult[0].browserCheckResult) {
    await respond(
      `Check ${checkId} is not a Browser check.\n<https://app.checklyhq.com/checks/${checkId}|Open the check in your Browser instead.>`,
    );
    return;
  }

  const errors = lastFailingCheckResult[0].browserCheckResult!.errors.map(
    (e) => e.stack,
  );

  const [prompt, config] = featureCoveragePrompt(
    check.name,
    check.scriptPath || "unknown",
    check.script || "unknown",
    errors,
  );
  const { text: summary } = await generateText({
    ...config,
    prompt,
  });

  await respond({
    text: summary,
  });
};
