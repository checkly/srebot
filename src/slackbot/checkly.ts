import { generateObject, generateText } from "ai";
import { checkly } from "../checkly/client";
import * as dataForge from "data-forge";
import {
  analyseCheckFailureHeatMap,
  summariseMultipleChecksGoal,
  summarizeErrorsPrompt,
  SummarizeErrorsPromptType,
  summarizeMultipleChecksStatus,
  summarizeTestGoalPrompt,
} from "../prompts/checkly";
import {
  fetchCheckResults,
  last24h,
  summarizeCheckResult,
} from "../prompts/checkly-data";
import { createCheckResultBlock } from "./blocks/checkResultBlock";
import { log } from "../log";
import { App, StringIndexed } from "@slack/bolt";
import { readCheck, readChecks } from "../db/check";
import { findErrorClustersForCheck } from "../db/error-cluster";
import {
  CheckResultAggregate,
  CheckResultTable,
  findCheckResults,
  findCheckResultsAggregated,
} from "../db/check-results";
import { readCheckGroup } from "../db/check-groups";
import generateCheckSummaryBlock from "./blocks/newCheckSummaryBlock";
import { analyseMultipleChecks } from "../use-cases/analyse-multiple/analyse-multiple-checks";
import { createMultipleCheckAnalysisBlock } from "./blocks/multipleChecksAnalysisBlock";
import { generateHeatmap } from "../heatmap/generateHeatmap";
import { createAccountSummaryBlock } from "./blocks/accountSummaryBlock";
import { aggregateCheckResults } from "./check-result-slices";
import { CheckResultsTimeSlice } from "./check-result-slices";

async function checkResultSummary(checkId: string, checkResultId: string) {
  const start = Date.now();
  const check = await checkly.getCheck(checkId);
  if (check.groupId) {
    const checkGroup = await checkly.getCheckGroup(check.groupId);
    check.locations = checkGroup.locations;
  }

  const checkAppUrl = checkly.getCheckUrl(check.id);
  const checkResult = await checkly.getCheckResult(check.id, checkResultId);
  const checkResultAppUrl = checkly.getCheckResultAppUrl(
    check.id,
    checkResult.id,
  );

  const interval = last24h(new Date(checkResult.startedAt));

  const checkResults = await fetchCheckResults(checkly, {
    checkId: check.id,
    ...interval,
  });

  const failingCheckResults = checkResults.filter(
    (result) => result.hasFailures || result.hasErrors,
  );

  const promptDef = summarizeErrorsPrompt({
    check: check.id,
    locations: check.locations,
    frequency: check.frequency,
    interval,
    results: [...failingCheckResults, checkResult].map(summarizeCheckResult),
  });
  const { object: errorGroups } =
    await generateObject<SummarizeErrorsPromptType>(promptDef);

  const heatmapImage = generateHeatmap(
    checkResults,
    interval.from,
    interval.to,
    {
      bucketSizeInMinutes: check.frequency * 10,
      verticalSeries: check.locations.length,
    },
  );

  log.info(
    {
      checkId,
      checkResultId,
      checkResultCount: checkResults.length,
      failingCheckResultCount: failingCheckResults.length,
      duration: Date.now() - start,
    },
    "checkResultSummary",
  );
  return {
    message: createCheckResultBlock({
      check,
      checkAppUrl,
      checkResult,
      checkResultAppUrl,
      errorGroups,
      failingCheckResults,
      intervalStart: interval.from,
    }),
    image: heatmapImage,
  };
}

async function checkSummary(checkId: string) {
  const start = Date.now();
  const check = await readCheck(checkId);
  if (check.groupId) {
    const checkGroup = await readCheckGroup(BigInt(check.groupId));
    check.locations = checkGroup.locations;
  }

  const prompt = summarizeTestGoalPrompt(
    check.name,
    check.script || "",
    check.scriptPath || "",
    [],
  );
  const { text: checkSummary } = await generateText(prompt);

  const interval = last24h(new Date());

  const { checkResults, checkCategory, heatmapImage } = await checkSummaryData(
    check.id,
    interval,
  );
  log.debug(
    {
      checkResultsLength: checkResults.length,
      durationMs: Date.now() - start,
      checkId,
    },
    "Fetched check results",
  );

  const failingCheckResults = checkResults.filter(
    (result) => result.hasFailures || result.hasErrors,
  );

  const failureClusters = await findErrorClustersForCheck(check.id);
  const failurePatterns = failureClusters.map((fc) => fc.error_message);

  const mostRecentFailureCheckResult = failingCheckResults.sort(
    (a, b) => new Date(b.startedAt).getTime() - new Date(a.startedAt).getTime(),
  )[0];
  const lastFailure =
    failingCheckResults.length > 0
      ? mostRecentFailureCheckResult.startedAt
      : checkResults[0].startedAt;

  const successRate = Math.round(
    ((checkResults.length - failingCheckResults.length) / checkResults.length) *
      100,
  );

  const heatmapPromptResult = await generateObject(
    analyseCheckFailureHeatMap(heatmapImage),
  );

  log.info(
    {
      checkId,
      checkResultCount: checkResults.length,
      failingCheckResultCount: failingCheckResults.length,
      duration: Date.now() - start,
    },
    "checkSummary",
  );
  const message = generateCheckSummaryBlock({
    checkId,
    checkName: check.name,
    checkSummary: checkSummary,
    checkState: checkCategory,
    lastFailure: new Date(lastFailure),
    successRate,
    failureCount: failingCheckResults.length,
    lastFailureId: mostRecentFailureCheckResult?.id,
    timeLocationSummary: heatmapPromptResult.object.failureIncidentsSummary,
    failurePatterns,
  });

  return { message, image: heatmapImage };
}

async function accountSummary(accountId: string) {
  const account = await checkly.getAccount(accountId);

  const interval = last24h(new Date());

  const statuses = await checkly.getStatuses();

  const counts = statuses.reduce(
    (acc, cr) => {
      if (!cr.hasErrors && !cr.hasFailures && !cr.isDegraded) {
        acc.passing++;
      }
      if (cr.isDegraded) {
        acc.degraded++;
      }
      if (cr.hasErrors || cr.hasFailures) {
        acc.failing++;
      }
      return acc;
    },
    { passing: 0, degraded: 0, failing: 0 },
  );

  log.info(
    {
      accountId,
      ...counts,
    },
    "accountSummary",
  );

  const aggregatedCheckResults = await findCheckResultsAggregated({
    accountId: accountId,
    from: interval.from,
    to: interval.to,
  });

  const labeledCheckResults = await summarizeCheckResultsToLabeledCheckStatus(
    aggregatedCheckResults,
  );

  const checkResultsWithCheckpoints = labeledCheckResults
    .toArray()
    .filter((cr) => cr.changePoints.length > 0);

  const { text: summary } = await generateText(
    summarizeMultipleChecksStatus(checkResultsWithCheckpoints),
  );

  const failingChecks = checkResultsWithCheckpoints.map((cr) => cr.checkId);
  const targetChecks = await readChecks(failingChecks);

  const { text: goals } = await generateText(
    summariseMultipleChecksGoal(targetChecks, 30),
  );

  const message = createAccountSummaryBlock({
    accountName: account.name,
    passingChecks: counts.passing,
    degradedChecks: counts.degraded,
    failingChecks: counts.failing,
    hasIssues: checkResultsWithCheckpoints.length > 0,
    issuesSummary: summary,
    failingChecksGoals: goals,
  });

  return { message };
}

function toTimeBucket(startedAt: Date, bucketSizeInMinutes: number) {
  const startedAtDate = new Date(startedAt);
  const minutes = startedAtDate.getMinutes();
  const sliceMinute =
    Math.floor(minutes / bucketSizeInMinutes) * bucketSizeInMinutes;
  const sliceDate = new Date(startedAtDate);
  sliceDate.setMinutes(sliceMinute, 0, 0); // Set to start of 30-min bucket
  return sliceDate;
}

const hourlyFormatter = new Intl.DateTimeFormat("en-US", {
  hour12: false,
  month: "2-digit",
  day: "2-digit",
  hour: "2-digit",
  minute: "2-digit",
  timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
});

export async function summarizeCheckResultsToLabeledCheckStatus(
  aggregatedCheckResults: CheckResultAggregate[],
) {
  let df = new dataForge.DataFrame(aggregatedCheckResults);

  const meanPassRatePerCheckAndLocation = df
    .groupBy((row) => `${row.checkId}|${row.runLocation}`)
    .select((group) => {
      const passing = group.deflate((row) => row.passingCount).sum();
      const degraded = group.deflate((row) => row.degradedCount).sum();
      const failing = group.deflate((row) => row.errorCount).sum();
      const passRateStdDev = group
        .deflate((row) => row.passingCount / row.count)
        .std();
      const meanPassRate = passing / (passing + degraded + failing);

      return {
        checkId: group.first().checkId,
        runLocation: group.first().runLocation,
        meanPassRate,
        passRateStdDev,
      };
    })
    .reduce((acc, row) => {
      acc.set(`${row.checkId}|${row.runLocation}`, {
        meanPassRate: row.meanPassRate,
        passRateStdDev: row.passRateStdDev,
      });
      return acc;
    }, new Map<string, { meanPassRate: number; passRateStdDev: number }>());

  log.debug("MEAN PASS RATE\n" + meanPassRatePerCheckAndLocation.toString());

  interface CheckTimeSlice {
    checkId: string;
    runLocation: string;
    startedAtBin: Date;
    passing: number;
    degraded: number;
    failing: number;
    passRate: number;
    passRateDiff: number;
    passRateStdDev: number;
    degradedRate: number;
    failRate: number;
    cumSumPassRate?: number;
  }

  const checkIdLocationTimeSliceWithPassRateStdDev = df
    .groupBy((row) => `${row.checkId}|${row.runLocation}|${row.startedAtBin}`)
    .select((group): CheckTimeSlice => {
      const checkId = group.first().checkId;
      const runLocation = group.first().runLocation;
      const { meanPassRate, passRateStdDev } =
        meanPassRatePerCheckAndLocation.get(`${checkId}|${runLocation}`)!;

      const count = group.deflate((row) => row.count).sum();
      const passing = group.deflate((row) => row.passingCount).sum();
      const degraded = group.deflate((row) => row.degradedCount).sum();
      const failing = group.deflate((row) => row.errorCount).sum();

      const passRate = passing / count;
      const degradedRate = degraded / count;
      const failRate = failing / count;

      return {
        checkId: group.first().checkId,
        runLocation: group.first().runLocation,
        startedAtBin: group.first().startedAtBin,
        passing,
        degraded,
        failing,
        passRate: passRate,
        passRateDiff: passRate - meanPassRate,
        passRateStdDev,
        degradedRate,
        failRate: failRate,
      };
    })
    .inflate()
    .withSeries<CheckTimeSlice & { cumSumPassRate: number }>(
      "cumSumPassRate",
      (s) => s.getSeries("passRateDiff").cumsum(),
    )
    .withSeries<
      CheckTimeSlice & { cumSumPassRate: number; isCheckPoint: boolean }
    >("isCheckPoint", (s) =>
      s.deflate((row) =>
        Math.abs(row.cumSumPassRate) > row.passRateStdDev * 2 ? 1 : 0,
      ),
    )
    .withSeries<
      CheckTimeSlice & {
        cumSumPassRate: number;
        isCheckPoint: boolean;
        checkPointGroup: number;
      }
    >("checkPointGroup", (s) =>
      new dataForge.Series([0]).concat(
        s
          .getSeries("isCheckPoint")
          .rollingWindow(2)
          .select((w) => {
            const firstIsCheckPoint = w.first();
            const lastIsCheckPoint = w.last();
            return firstIsCheckPoint === lastIsCheckPoint ? 0 : 1;
          })
          .cumsum(),
      ),
    );

  log.debug(
    "CHECK ID LOCATION TIME SLICE WITH PASS RATE STD DEV\n" +
      checkIdLocationTimeSliceWithPassRateStdDev.toString(),
  );

  const checksWithChangePoints = checkIdLocationTimeSliceWithPassRateStdDev
    .groupBy((row) => `${row.checkId}|${row.runLocation}`)
    .select((group) => {
      const changePoints = group
        .filter((row) => row.isCheckPoint)
        .groupBy((row) => row.checkPointGroup)
        .select((group) =>
          group.orderBy((row) => Math.abs(row.cumSumPassRate)).last(),
        )
        .toArray();

      return {
        checkId: group.first().checkId,
        runLocation: group.first().runLocation,
        changePoints: changePoints.map((cp) => ({
          timestamp: cp.startedAtBin.getTime(),
          formattedTimestamp: hourlyFormatter.format(cp.startedAtBin),
          severity:
            cp.failRate > 0
              ? "FAILING"
              : cp.degradedRate > 0
                ? "DEGRADED"
                : "PASSING",
        })),
      };
    })
    .inflate();

  log.debug("CHECKS WITH CHANGE POINTS\n" + checksWithChangePoints.toString());

  return checksWithChangePoints;
}

async function checkSummaryData(
  checkId: string,
  interval: { from: Date; to: Date },
) {
  const checkResults = await findCheckResults(
    checkId,
    interval.from,
    interval.to,
  );

  if (checkResults.length === 0) {
    return {
      checkId,
      checkResults: [] as CheckResultTable[],
      runLocations: new Set<string>(),
      checkCategory: "PASSING",
      heatmapImage: Buffer.from([]),
      lastRun: null,
      lastFailure: null,
      status: "passing",
      timeSlices: [] as CheckResultsTimeSlice[],
    };
  }

  const timeSlices = aggregateCheckResults(
    checkResults,
    interval.from,
    interval.to,
  );

  const lastRun = checkResults[0];
  const lastFailure = checkResults.find((cr) => cr.hasFailures || cr.hasErrors);

  const status =
    lastRun.hasFailures || lastRun.hasErrors
      ? "failing"
      : lastRun.isDegraded
        ? "degraded"
        : "passing";

  const runLocations = checkResults.reduce((acc, cr) => {
    acc.add(cr.runLocation);
    return acc;
  }, new Set<string>());

  const heatmapImage = generateHeatmap(
    checkResults,
    interval.from,
    interval.to,
    {
      bucketSizeInMinutes: 30,
      verticalSeries: runLocations.size,
    },
  );

  const checkCategory = (
    await generateObject(analyseCheckFailureHeatMap(heatmapImage))
  ).object.category;

  return {
    checkId,
    checkResults,
    runLocations,
    checkCategory,
    heatmapImage,
    lastRun,
    lastFailure,
    status,
    timeSlices,
  };
}

// Allow overriding the command name for local dev
export const CHECKLY_COMMAND_NAME =
  process.env.CHECKLY_COMMAND_NAME_OVERRIDE || "/checkly";

const getIsUUID = (str: string): boolean => {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(
    str,
  );
};

export const checklyCommandHandler = (app: App<StringIndexed>) => {
  return async ({ ack, respond, command }) => {
    await ack();
    const args = command.text.split(" ");
    if (args.length == 1 && args[0].trim() === "") {
      const accountId = process.env.CHECKLY_ACCOUNT_ID!;
      const { message } = await accountSummary(accountId);
      await respond(message);
    } else if (args.length == 1 && !getIsUUID(args[0])) {
      const multipleCheckAnalysisResult = await analyseMultipleChecks(args[0]);
      await respond({
        ...createMultipleCheckAnalysisBlock(multipleCheckAnalysisResult),
        response_type: "in_channel",
      });
    } else if (args.length === 1 && !!args[0] && getIsUUID(args[0])) {
      const { message, image } = await checkSummary(args[0]);
      await respond({
        response_type: "in_channel",
        ...message,
      });

      if (image) {
        await app.client.files.uploadV2({
          channel_id: command.channel_id,
          file: image,
          filename: "CheckResultHeatmap.png",
          title: "Check Results Heatmap",
        });
      }

      // FIXME find a way to send the image to slack (al)
    } else if (args.length === 2) {
      const [checkId, checkResultId] = args;
      const { message, image } = await checkResultSummary(
        checkId,
        checkResultId,
      );

      if (image) {
        await app.client.files.uploadV2({
          channel_id: command.channel_id,
          file: image,
          filename: "CheckResultHeatmap.png",
          title: "Check Results Heatmap",
        });
      }

      await respond({
        response_type: "in_channel",
        ...message,
      });
    } else {
      await respond({
        text: "Please provide either a check ID or both a check ID and check result ID in the format: /checkly <check_id> (<check_result_id>)",
      });
    }
  };
};
