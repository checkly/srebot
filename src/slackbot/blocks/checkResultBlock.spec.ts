import { describe, expect, it } from "@jest/globals";
import { createCheckResultBlock } from "./checkResultBlock";
import { ChecklyClient } from "../../checkly/checklyclient";
import { Check, CheckResult } from "../../checkly/models";

describe("checkResultBlock", () => {
  it("should create a message block for a check result", () => {
    const mockCheck: Check = {
      id: "test-check-id",
      name: "Test Check",
      locations: ["us-east-1", "eu-west-1"],
    } as Check;

    const mockCheckResult: CheckResult = {
      id: "test-result-id",
      runLocation: "us-east-1",
    } as CheckResult;

    const mockErrorGroups = {
      groups: [
        {
          errorMessage: "Test error message",
          checkResults: ["test-result-id"],
        },
      ],
    };

    const result = createCheckResultBlock({
      check: mockCheck,
      checkAppUrl: "https://app.checklyhq.com/checks/test-check-id",
      checkResult: mockCheckResult,
      checkResultAppUrl:
        "https://app.checklyhq.com/checks/test-check-id/results/test-result-id",
      errorGroups: mockErrorGroups,
      failingCheckResults: [mockCheckResult],
      intervalStart: new Date("2024-01-01T00:00:00.000Z"),
    });

    expect(result.blocks).toMatchSnapshot();
  });
});
