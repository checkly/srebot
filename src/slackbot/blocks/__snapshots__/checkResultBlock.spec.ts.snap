// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkResultBlock should create a message block for a check result 1`] = `
[
  {
    "text": {
      "emoji": true,
      "text": "Check Result Details",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "fields": [
      {
        "text": "*Check Name:*
<https://app.checklyhq.com/checks/test-check-id|Test Check>",
        "type": "mrkdwn",
      },
      {
        "text": "*Timestamp:*
2024-01-01T00:00:00.000Z",
        "type": "mrkdwn",
      },
      {
        "text": "*Location:*
\`us-east-1\`",
        "type": "mrkdwn",
      },
      {
        "text": "*Check Result:*
<https://app.checklyhq.com/checks/test-check-id/results/test-result-id|Link>",
        "type": "mrkdwn",
      },
    ],
    "type": "section",
  },
  {
    "text": {
      "text": "*Error Message:*
\`Test error message\`

*Similar Failures:*
This error occurred \`1 times\` across locations: \`us-east-1\`, \`eu-west-1\`. We saw \`1 Error Groups\` in total with \`1 failures\` in the same timeframe.",
      "type": "mrkdwn",
    },
    "type": "section",
  },
]
`;
